### PR TITLE
feat: add count_rows to scanner

### LINKED ISSUE #841: Support `count_rows()` with filter expression.
# Problem Statement

`Dataset.count_rows()` does quick / cheap calculation of total rows in the dataset. We have not yet implemented the `dataset.count_rows(expression)`. 

# Solution

Support `Dataset::count_rows()` with filters, with minor overheads (i.e., only reads the necessary columns in the filter)

### PR DESCRIPTION
**Pull request by westonpace**
Adds a `count_rows` method to the scanner which will take any existing scanner filter into account.

Closes #841

### CODE REVIEW

**Comment by westonpace** [comment]
#841 asks for a change to the `Dataset` method.  I can do that if we want.  However, there are two small advantages to the current approach.

 * Avoids a breaking change - pretty minor advantage since the API is unstable?
 * This way is more obvious to the user that they will be paying the cost of a scan

In other words, currently, one must do `dataset.scan().filter(...)?.count_rows()` instead of `dataset.count_rows(Some(...))`

**Comment by westonpace** [comment]
Leaving in draft as this PR is using some of the datagen utilities from #1140

File: rust/src/dataset/scanner.rs [hunk 1]
```diff
@@ -350,20 +354,63 @@ impl Scanner {
...
+    }
+
+    /// Create a stream from the Scanner.
+    pub async fn try_into_stream(&self) -> Result<DatasetRecordBatchStream> {
+        let plan = self.create_plan().await?;
+        Ok(DatasetRecordBatchStream::new(self.execute_plan(plan)?))
+    }
+
+    /// Scan and return the number of matching rows
+    pub async fn count_rows(&self) -> Result<usize> {
```
Comment (eddyxu) [inline rust/src/dataset/scanner.rs:375]: is there plan to support `count_rows(filter)` ?

Comment (westonpace) [reply rust/src/dataset/scanner.rs:375]: This `count_rows` method is on the scanner, which may have a filter already specified.  So you can do:

```
            dataset
                .scan()
                .filter("filter_me > 15")
                .unwrap()
                .count_rows()
```

**Review by eddyxu — APPROVED** [review]
I see. Make sense.

File: rust/src/dataset/scanner.rs [hunk 2]
```diff
@@ -350,20 +354,63 @@ impl Scanner {
...
+            vec![None],
+            plan,
+            plan_schema,
+        )?);
+        let mut stream = self.execute_plan(count_plan)?;
+
+        // A count plan will always return a single batch with a single row.
+        if let Some(first_batch) = stream.next().await {
+            let batch = first_batch?;
+            dbg!(&batch);
```
Comment (wjones127) [inline rust/src/dataset/scanner.rs:402]:
```suggestion
```

**Review by wjones127 — COMMENTED** [review]
@westonpace just one cleanup task then looks good. I assume Python bindings are a follow up?

File: rust/src/dataset/scanner.rs (hunk 2 above)
Comment (westonpace) [reply rust/src/dataset/scanner.rs]: Thanks :)  I've removed this line.

**Comment by westonpace** [comment]
> 

I went ahead and added python bindings

**Pull request merged**